code/__HELPERS/traits.dm

```diff
@@ -2,89 +2,116 @@
 #define SIGNAL_REMOVETRAIT(trait_ref) "removetrait [trait_ref]"
 
 // trait accessor defines
+
+/**
+ * Adds a status trait to the target datum.
+ *
+ * Arguments: (All Required)
+ * * target - The datum to add the trait to.
+ * * trait - The trait which is being added.
+ * * source - The source of the trait which is being added.
+ */
 #define ADD_TRAIT(target, trait, source) \
 	do { \
-		var/list/_L; \
-		if (!target.status_traits) { \
-			target.status_traits = list(); \
-			_L = target.status_traits; \
-			_L[trait] = list(source); \
-			SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
+		LAZYINITLIST(target.status_traits); \
+\
+		if(!target.status_traits[trait]) { \
+			target.status_traits[trait] = list(source); \
 		} else { \
-			_L = target.status_traits; \
-			if (_L[trait]) { \
-				_L[trait] |= list(source); \
-			} else { \
-				_L[trait] = list(source); \
-				SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
-			} \
+			target.status_traits[trait] |= list(source); \
 		} \
+\
+		SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
 	} while (0)
+
+/**
+ * Removes a status trait from a target datum.
+ *
+ * `ROUNDSTART_TRAIT` traits can't be removed without being specified in `sources`.
+ * Arguments:
+ * * target - The datum to remove the trait from.
+ * * trait - The trait which is being removed.
+ * * sources - If specified, only remove the trait if it is from this source. (Lists Supported)
+ */
 #define REMOVE_TRAIT(target, trait, sources) \
 	do { \
-		var/list/_L = target.status_traits; \
-		var/list/_S; \
-		if (sources && !islist(sources)) { \
-			_S = list(sources); \
-		} else { \
-			_S = sources\
-		}; \
-		if (_L && _L[trait]) { \
-			for (var/_T in _L[trait]) { \
-				if ((!_S && (_T != ROUNDSTART_TRAIT)) || (_T in _S)) { \
-					_L[trait] -= _T \
+		if(target.status_traits && target.status_traits[trait]) { \
+			var/list/TARGET_TRAITS = target.status_traits; \
+			var/list/SOURCES = sources; \
+			if(sources && !islist(sources)) { \
+				SOURCES = list(sources); \
+			} \
+\
+			for(var/SOURCE in TARGET_TRAITS[trait]) { \
+				if((!SOURCES && (SOURCE != ROUNDSTART_TRAIT)) || (SOURCE in SOURCES)) { \
+					if(length(TARGET_TRAITS[trait]) == 1) { \
+						SEND_SIGNAL(target, SIGNAL_REMOVETRAIT(trait), trait); \
+					} \
+					LAZYREMOVEASSOC(TARGET_TRAITS, trait, SOURCE); \
```

Comment:
If I'm reading this right, you're now checking for length == 1 instead because the `LAZYREMOVEASSOC` is about to remove the last value right after this `if`.

If that's true I think it would be better to put the removal before the `if` and keep the 0 length check like how it was before.
```dm
LAZYREMOVEASSOC(TARGET_TRAITS, trait, SOURCE); \
if(!length(TARGET_TRAITS[trait])) { \
	SEND_SIGNAL(target, SIGNAL_REMOVETRAIT(trait), trait); \
} \
```

Reply:
Since `LAZYREMOVEASSOC()` also nulls the list entirely if it removes the last enty, `if(!length(null[trait]))` will give a "bad index" runtime.
This could be avoided by writing out `LAZYREMOVEASSOC()` manually like it was before, but I felt that using the existing define for readability was worth the seemingly arbitrary length check.
![image](https://user-images.githubusercontent.com/57483089/126496577-f7573f8a-5d06-4809-ae7f-bbd779e83b93.png)